tests/run-macros/i11685/Macro_1.scala

```diff
@@ -0,0 +1,79 @@
+package test
+
+import scala.quoted.*
+
+trait Thing {
+  type Type
+}
+
+object MyMacro {
+
+  def isExpectedReturnType[R: Type](using Quotes): quotes.reflect.Symbol => Boolean = { method =>
+    import quotes.reflect.*
+
+    val expectedReturnType = TypeRepr.of[R]
+
+    method.tree match {
+      case DefDef(_,_,typedTree,_) =>
+          TypeRepr.of(using typedTree.tpe.asType) <:< expectedReturnType
+      case _ => false
+    }
+  }
+
+  ///TODO no overloads
+  def checkMethod[R: Type](using q: Quotes)(method: quotes.reflect.Symbol): Option[String] = {
+    val isExpectedReturnTypeFun = isExpectedReturnType[R]
+
+    Option.when(method.paramSymss.headOption.exists(_.exists(_.isType)))(s"Method ${method.name} has a generic type parameter, this is not supported") orElse
+      Option.when(!isExpectedReturnTypeFun(method))(s"Method ${method.name} has unexpected return type")
+  }
+
+  def definedMethodsInType[T: Type](using Quotes): List[quotes.reflect.Symbol] = {
+    import quotes.reflect.*
+
+    val tree = TypeTree.of[T]
+
+    for {
+      member <- tree.symbol.methodMembers
+      //is abstract method, not implemented
+      if member.flags.is(Flags.Deferred)
+
+      //TODO: is that public?
+      // TODO? if member.privateWithin
+      if !member.flags.is(Flags.Private)
+      if !member.flags.is(Flags.Protected)
+      if !member.flags.is(Flags.PrivateLocal)
+
+      if !member.isClassConstructor
+      if !member.flags.is(Flags.Synthetic)
+    } yield member
+  }
+
+  transparent inline def client[T, R](r: () => R): T = ${MyMacro.clientImpl[T, R]('r)}
+
+  def clientImpl[T: Type, R: Type](r: Expr[() => R])(using Quotes): Expr[T] = {
+    import quotes.reflect.*
+
+    val apiType = TypeRepr.of[T]
+    val tree = TypeTree.of[T]
+
+    val methods = definedMethodsInType[T]
+    val invalidMethods = methods.flatMap(checkMethod[R])
+    if (invalidMethods.nonEmpty) {
+      report.errorAndAbort(s"Invalid methods: ${invalidMethods.mkString(", ")}")
+    }
+
+    val className = "_Anon"
+    val parents = List(TypeTree.of[Object], TypeTree.of[T])
+
+    def decls(cls: Symbol): List[Symbol] = methods.map { method =>
+      Symbol.newMethod(cls, method.name, method.info, flags = Flags.EmptyFlags /*TODO: method.flags */, privateWithin = method.privateWithin.fold(Symbol.noSymbol)(_.typeSymbol))
```

Comment:
I see you opened https://github.com/lampepfl/dotty/issues/15035 but it's classified as an enhancement when I think it should be required to stabilize Symbol.newClass

Reply:
I'm working on that one now. I intend to open a separate PR for that one as it is not directly related with the `info`.

Reply:
See #15041